### PR TITLE
[SPARK-34692][SQL] Support Not(In) and Not(InSet) propagate null in predicate

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

* Add `Not(In)` and `Not(InSet)` check in `NullPropagation` rule.
* Add more test for `In` and `Not(In)` in `Project` level.

### Why are the changes needed?

The semantics of `Not(In)` could be seen like `And(a != b, a != c)` that match the `NullIntolerant`.

As we already simplify the `NullIntolerant` expression to null if it's children have null. E.g. `a != null` => `null`. It's safe to do this with `Not(In)`/`Not(InSet)`.

Note that, we can only do the simplify in predicate which `ReplaceNullWithFalseInPredicate`  rule do.

Let's say we have two sqls:
```
select 1 not in (2, null);
select 1 where 1 not in (2, null);
```
The first sql we cannot optimize since it would return `NULL` instead of `false`. The second is postive.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Add test.

### CODE REVIEW

**Comment by tanelk** [comment]
Are you sure, that this rule is correct.
For example `Not(1 IN (1, NULL))` would be `Not(true)` or `false`.
But with this rule it would turn into `NULL`

**Comment by tanelk** [comment]
The `NullPropagation` applies to all expressions, but `ReplaceNullWithFalseInPredicate` changes only `Filter`, `Join` and some others. So if I would write the results of `Not(1 IN (1, NULL))` to file, then this rule would change it from `false` to `null`.

**Comment by ulysses-you** [comment]
ah, you are right. We should check the plan level at least otherwise it's not safe.

Seems we missed such test case you said, I think it's worth to add some test for it.

**Comment by tanelk** [comment]
I'm not 100% sure, but perhaps these rules would fit in `ReplaceNullWithFalseInPredicate.replaceNullWithFalse(e: Expression)`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceNullWithFalseInPredicate.scala [hunk 1]
```diff
@@ -95,6 +95,11 @@ object ReplaceNullWithFalseInPredicate extends Rule[LogicalPlan] {
   private def replaceNullWithFalse(e: Expression): Expression = e match {
     case Literal(null, BooleanType) =>
       FalseLiteral
+    case Not(In(value, list)) if (value +: list).exists(isNullLiteral) =>
```
Comment (ulysses-you) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceNullWithFalseInPredicate.scala:98]: `null in (xxx)`  is a special case that we have already handled in `NullPropagation`, but it's better that return the same result whether other rule is excluded. We also add the `value` pattern here.

**Comment by ulysses-you** [comment]
cc @tanelk @maropu @cloud-fan do you have time to take a look at this, it would be so kind if you can give any suggestion.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceNullWithFalseInPredicate.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceNullWithFalseInPredicate.scala:98]: Can we add some comments to explain the rationale?

if `value` is null, the IN result is null and thus `Not(null)` means false. If `value` is not null but the list contains null values, the IN result can either be `true` or `null`, and thus the final result after Not is always false as a predicate.

Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/ReplaceNullWithFalseInPredicate.scala:104]: Yes, it's better to add the comment.

File: sql/core/src/test/resources/sql-tests/inputs/predicate-functions.sql [hunk 2]
```diff
@@ -46,3 +46,23 @@ select to_timestamp('2017-03-01 00:00:01') > to_date('2017-03-01');
...
+
+-- Not(In)
+select 1 not in (1, 2, 3);
+select 1 not in (1, 2, 3, null);
+select 1 not in (1.0, 2.0, 3.0);
+select 1 not in (1.0, 2.0, 3.0, null);
+select 1 not in ('2', '3', '4');
+select 1 not in ('2', '3', '4', null);
+select null not in (1, 2, 3);
+select null not in (1, 2, null);
```
Comment (ulysses-you) [inline sql/core/src/test/resources/sql-tests/inputs/predicate-functions.sql:68]: These test is for better coverage and avoid the regression from this PR.

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
retest this please

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**